Control/lib/control-core/EnvCache.js

```diff
@@ -0,0 +1,49 @@
+/**
+ * @license
+ * Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+ * See http://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+ * All rights not expressly granted are reserved.
+ *
+ * This software is distributed under the terms of the GNU General Public
+ * License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+ *
+ * In applying this license CERN does not waive the privileges and immunities
+ * granted to it by virtue of its status as an Intergovernmental Organization
+ * or submit itself to any jurisdiction.
+*/
+
+/**
+ * Caches AliECS core GetEnvironments response
+ */
+class EnvCache {
+
+  /**
+   * @param {object} ctrlService - Handle to Control service
+   */
+  constructor(ctrlService) {
+    this.ctrlService = ctrlService;
+    this.cache = {};
+    this.refreshInterval = setInterval(() => this.refresh(), 1000);
```

Comment:
I still believe we should check if previous request completed. 
There is no point in flooding AliECS with hundreds of requests if it is hanging. It will only come back to us as request to address it.

Reply:
Okay, @teo what would be reasonable time to wait for `GetEnvironments` to complete?

Reply:
@graduta Indeed, I'm also not sure what to do when `GetEnvironments` timeouts? Do we clear envs from cache? Maybe it's better to keep them but clear the list only on AliECS core restart (is there a way to detect it?).

Reply:
Hi @awegrzyn . From my point of view I would propose we:
1. keep the list so that shifters can easily copy-paste the env id so that they search in infologger
2. Display an informative message to the user that AliECS hangs
3. We do not allow the user to go to "More"of an environment as AliECS is hanging anyway
4. If AliECS is back then we get a new list of Envs and Cache is updated

gPRC.js will send back a standard "DEADLINE_EXCEEDED" message if timeout is reached but other than that we cannot tell if something is wrong. As we also have lengthy times for creating an environment, I would say is not ok to assume anything until we get the timeout message

Reply:
> Okay, @teo what would be reasonable time to wait for `GetEnvironments` to complete?

Not sure, a few seconds? It should reply immediately or almost.